### PR TITLE
Got rid of Axon* interfaces, added new erand.SysRand random generator to Network, use for InitWts etc.

### PR DESCRIPTION
**Pull request by rcoreilly**
projects mostly require replacing `emer.Forward` -> `axon.ForwardPrjn` etc for other emer.* types, in Connect methods -- otherwise it is fairly transparent, and main axon code is much cleaner -- no more casting all the time!

### CODE REVIEW

**Comment by rcoreilly** [comment]
needs more fixing of random number stuff in various places, and testing, impl in examples.

**Review by siboehm — COMMENTED** [review]
Uff just spent >1h on this and am still not done. This PR is almost impossible to review. It should've been 3 PRs: 1 for inconsequential renamings, 2 for the randomness, 3 for getting rid of the interface. That way, I could've approved 1 + 3 after a quick glance, and given 2 proper thought. As a single PR I have to go through 2000 lines of changes to find the impactful stuff (the randomness). Tried going through it by commit instead, but that wasn't much more helpful. Any chance you can split this into 3 PRs, or describe in more detail what & where you're changing the randomness?

Comments:
1. Any reason why you didn't fully get rid of the interfaces?
1. I don't like the LayByName refactor since it's less descriptive and since Lay isn't used as a shortform for Layer anywhere else in the codebase. But I like discussions about naming even less, so keep it as is if you like.

File: axon/act.go [hunk 1]
```diff
@@ -150,10 +149,10 @@ func (ai *ActInitParams) Defaults() {
 //gosl: end act
 
 // GeBase returns the baseline Ge value: Ge + rand(GeVar) > 0
-func (ai *ActInitParams) GetGeBase() float32 {
+func (ai *ActInitParams) GetGeBase(nt *Network) float32 {
```
[the remark this answers is not in the document]
Comment (rcoreilly) [reply axon/act.go:152]: good idea -- will change that.

File: axon/axon.go [hunk 2]
```diff
@@ -34,35 +33,6 @@ type AxonNetwork interface {
 	// AxonNetwork interface does not need to include accessors
 	// to all the basic stuff
 	AsAxon() *Network
```
[the remark this answers is not in the document]
Comment (rcoreilly) [reply axon/axon.go:35]: It allows existing projects to build without switching over to non-interface version, and can be useful if we need something in NetworkBase to be able to call Network method (we do this in Layer).  Anyway we can do this later once we're fully sure we don't need it.

File: axon/networkbase.go [hunk 3]
```diff
@@ -106,9 +108,22 @@ func (nt *NetworkBase) LayerByNameTry(name string) (emer.Layer, error) {
 	return ly, nil
 }
 
+// LayByName returns a layer by looking it up by name in the layer map (nil if not found).
+// Will create the layer map if it is nil or a different size than layers slice,
+// but otherwise needs to be updated manually.
+func (nt *NetworkBase) LayerByName(name string) emer.Layer {
```
[the remark this answers is not in the document]
Comment (rcoreilly) [reply axon/networkbase.go:114]: Yeah this isn't great -- LMK if you have a better plan:
* `LayerByName` is the `emer.Network` interface and is necessary for emer infrastructure -- returns emer version
* `LayByName` returns the *axon.Layer

**Comment by rcoreilly** [comment]
it is just adding the local random seed and using it for all random calls..

File: axon/networkbase.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (siboehm) [reply axon/networkbase.go:114]: I'd make my life easy and Go `AxonLayerByName` and `LayerByName` for the emer version. Could also switch them around to make less typing, but I assume you don't want to change the emer interface.

In any case, the comment just be adjusted to make it clear why there's two different functions for (almost) the same thing.

[the remark this answers is not in the document]
Comment (rcoreilly) [reply axon/networkbase.go:114]: yeah AxonLayerByName is good -- can't change the `emer` interface..

**Review by siboehm — APPROVED** [review]

**Pull request merged**